python/mxnet/contrib/nn/layer.py

```diff
@@ -0,0 +1,319 @@
+        to prevent name collisions.
+    params : ParameterDict or None
+        Manages Parameters of this Layer and sublayers. You can make two Layers share
+        parameter by passing the same dictionary to them. For example::
+            params = nn.ParameterDict(prefix='dense_')
+            dense1 = nn.Dense(20, in_units=10, prefix='dense1_', params=params)
+            dense2 = nn.Dense(20, in_units=10, prefix='dense2_', params=params)
+
+        dense1 and dense2 now have shared weights.
+    """
+    _counter = {}
+
+    def __init__(self, prefix=None, params=None):
+        self._prefix = self._make_prefix(prefix)
+        self._name = self.prefix[:-1] if self.prefix.endswith('_') else self.prefix
+        if params is None:
+            params = ParameterDict(prefix=self.prefix)
+        self._params = params
+        self._children = []
+
+    def __setattr__(self, name, value):
+        """Automatically register sublayers."""
+        if isinstance(value, Layer):
+            self.register_sublayer(value)
+        super(Layer, self).__setattr__(name, value)
+
+    def _make_prefix(self, prefix=None):
+        """Create prefix. By default use class name plus counter."""
+        if prefix is None:
+            count = Layer._counter.get(type(self), 0)
+            Layer._counter[type(self)] = count + 1
+            return '%s%d_'%(self.__class__.__name__.lower(), count)
+        return prefix
+
+    @property
+    def params(self):
+        """A ParameterDict managing this Layer's Parameters."""
+        return self._params
+
+    @property
+    def prefix(self):
+        """Prefix of this Layer."""
+        return self._prefix
+
+    def register_sublayer(self, layer):
+        """Register layer as sublayer of self. Layers assigned to self as attributes
+        will be registered automatically."""
+        self._children.append(layer)
+        self.params.merge(layer.params)
+
+    def __call__(self, *args, **kwargs):
+        """Call forward."""
+        return self.forward(*args, **kwargs)
+
+    def forward(self, *args, **kwargs):
+        """Defines the forward computation. Arguments can be either NDArray or Symbol."""
+        raise NotImplementedError
+
+
+class Sequential(Layer):
+    """Stack Layers sequentially.
+
+    Example::
+        net = nn.Sequential()
+        net.add(Dense(10, activation='relu'))
+        net.add(Dense(20))
+    """
+    def __init__(self, params=None, **kwargs):
+        super(Sequential, self).__init__(prefix='', params=params, **kwargs)
+
+    def add(self, layer):
+        """Add layer on top of the stack."""
+        self.register_sublayer(layer)
+
+    def forward(self, x):
+        #pylint: disable=arguments-differ
+        for layer in self._children:
+            x = layer(x)
+        return x
+
+
+class SimpleLayer(Layer):
+    """SimpleLayer is a Layer that supports forwarding with both `Symbol` and `NDArray`.
```

Comment:
What's the difference between SimpleLayer and Layer? Layer cannot forward with NDArray?